### PR TITLE
The beginnings of a PSV test

### PR DESCRIPTION
**Pull request by jystic**
This also fixes a bunch of bugs caught by the test.

This borrowed a bunch of stuff from the seaworthy test so it needs to be updated to use the new pipeline.

The next step is actually feeding it the generated input data as a PSV file.

### CODE REVIEW

File: src/Icicle/Sea/FromAvalanche/Analysis.hs [hunk 1]
```diff
@@ -168,3 +173,61 @@ outputsOfStatement stmt
...
+     Write _ x             -> typesOfExp x
+     LoadResumable _ vt    -> Set.singleton vt
+     SaveResumable _ vt    -> Set.singleton vt
+     Output _ vt xs        -> Set.singleton vt `Set.union`
+                              Set.unions (fmap goOut xs)
+     KeepFactInHistory     -> Set.empty
+
+     Read _ _ vt ss
+      -> Set.singleton vt `Set.union`
+         typesOfStatement ss
```
Comment (tranma) [inline src/Icicle/Sea/FromAvalanche/Analysis.hs]: I wonder if building up a list of types then just do Set.fromList at the end is faster. It's not a big deal though, perf is not bad atm.

**Comment by tranma** [comment]
yep, WellTyped sounds good, just need to use pipeline stuff for it :+1:

File: src/Icicle/Sea/FromAvalanche/Analysis.hs (hunk 1 above)
Comment (jystic) [inline src/Icicle/Sea/FromAvalanche/Analysis.hs]: I'm not sure, I plucked this from the C++ branch.

I suspect you'd end up doing the same number of comparisons either way right? I'm not sure.

Comment (amosr) [inline src/Icicle/Sea/FromAvalanche/Analysis.hs]: I'd be curious to see - I have no real intuition about why fromList would be faster

File: data/sea/20-simple.h [hunk 2]
```diff
@@ -57,8 +57,9 @@ static ibool_t   INLINE istring_ne    (istring_t x, istring_t y) { return strcmp
 
 static istring_t INLINE istring_copy(imempool_t *into, istring_t val)
 {
-    char* alloc = imempool_alloc(into, strlen(val));
-    strcpy(alloc, val);
+    size_t val0_size = strlen (val) + 1;
+    char  *alloc     = imempool_alloc (into, val0_size);
+    memcpy (alloc, val, val0_size);
```
Comment (amosr) [inline data/sea/20-simple.h]: oh crap. good catch

File: src/Icicle/Sea/FromAvalanche/Psv.hs [hunk 3]
```diff
@@ -242,26 +239,68 @@ checkInputType state
...
+        [ ""
+        , "if (value_size == sizeof (\"true\") - 1 &&"
+        , "    memcmp (\"true\", value_ptr, value_size)) {"
+        , "    program->" <> pretty nx <> "[program->new_count] = itrue;"
+        , "} else if (value_size == sizeof (\"false\") - 1 &&"
+        , "           memcmp (\"false\", value_ptr, value_size)) {"
+        , "    program->" <> pretty nx <> "[program->new_count] = ifalse;"
+        , "} else {"
+        , "    return psv_alloc_error (\"not a boolean\", value_ptr, value_size);"
+        , "}"
```
Comment (amosr) [inline src/Icicle/Sea/FromAvalanche/Psv.hs]: would these be nicer in a function in the preamble?

**Comment by amosr** [comment]
:+1:

File: src/Icicle/Sea/FromAvalanche/Analysis.hs (hunk 1 above)
Comment (jystic) [inline src/Icicle/Sea/FromAvalanche/Analysis.hs]: I hope this is just as fast, I use this pattern all the time, containers yields such beautiful code in this arrangement :)

Comment (tranma) [inline src/Icicle/Sea/FromAvalanche/Analysis.hs]: yeah. I was wondering if there would be fewer comparisons if you build up the set afterwards, but it's probably best to leave the performance fiddling to containers :)

**Pull request merged**